drivers: spi: make SPI dt-spec macros compatible with C++

gmarull opened the pull request:
As of today it is not possible to use SPI dt-spec macros in C++, something known and documented. The main reason is because `cs` property is initialized using a compound literal, something not supported in C++. This PR takes another approach, that is to not make `cs` a pointer but a struct member. This way, we can perform a regular initialization, at the cost of using extra memory for unused delay/pin/flags if `cs` is not used.

Fixes #56572
Closes #57025

gmarull committed:
drivers: spi: make SPI dt-spec macros compatible with C++

As of today it is not possible to use SPI dt-spec macros in C++,
something known and documented. The main reason is because `cs` property
is initialized using a compound literal, something not supported in C++.
This PR takes another approach, that is to not make `cs` a pointer but a
struct member. This way, we can perform a regular initialization, at the
cost of using extra memory for unused delay/pin/flags if `cs` is not
used.

Fixes #56572

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
tests: mec15xxevb_assy6853: qspi: make module vars static

Make some module variables static, limiting their scope and guaranteeing
zero-initialization. This allows to remove some redundant code as well.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

cfriedt reviewed (dismissed):
LGTM - just wondering if this falls under the category of a breaking / tree-wide change and if it needs to go through some process.

gmarull commented:
> LGTM - just wondering if this falls under the category of a breaking / tree-wide change and if it needs to go through some process.

Hmm I'm not sure if it applies, since the changes are restricted to something that applications should not use (we should actually isolate internals better here).

teburd commented on include/zephyr/drivers/spi.h:
```diff
@@ -319,10 +304,9 @@ struct spi_config {
 #endif /* CONFIG_SPI_EXTENDED_MODES */
 
-	const struct spi_cs_control *cs;
+	struct spi_cs_control cs;
```
This is a break API change and needs to go through the process. If an out of tree driver implements and uses spi_config today, and goes to update zephyr, this will break that code.

I'm 100% fine with this case myself (though @tbursztyka will need to +1 it). But it needs to be announced on the mailing list and at least be brought up in the architecture meeting.

gmarull replied:
If we go this road, I'd also improve isolation of this module. Applications should, in most cases, not play with `cs` field directly. Some helper function would have been useful to prevent a breaking change here.

teburd replied:
Unfortunately spi_config is a public struct in a public header, so even with helpers there's no guarantees. It's something we need to live with and work with. The process here is really about, making sure the community is ok with it and getting the word out.

That's ok, but does mean a quick blurb in the architecture meeting and an email. Easy

galak reviewed (commented):
Can this be done as a deprecation of the old "interface" and intro a new Kconfig for the new means and have the various in-trees select that new kconfig?

cfriedt commented on include/zephyr/drivers/spi.h:
```diff
@@ -229,39 +229,27 @@ struct spi_cs_control {
  * @return a pointer to the @p spi_cs_control structure
  */
-#define SPI_CS_CONTROL_PTR_DT(node_id, delay_)			  \
-	(&(struct spi_cs_control) {				  \
```
Just caught this with the arch meeting and wanted to mention it here too.

The old spi macro declaration appears to be a statement expr(ession). It's a gcc-ism (just happens to be supported by clang too), but they aren't ISO C or C++. We (somewhat unfortunately) do have them all over the place still.

I would say that the preferred way is to embed the struct, which is what Gerrard has done. We pay the price in terms of an extra 4 bytes, but it's at least ISO C / C++.

gmarull committed:
drivers: spi: add spi_cs_is_gpio(_dt) helpers

Add spi_cs_is_gpio(_dt) helpers to check whether SPI CS is controlled by
GPIO or not. This both improves code readability and isolates SPI
internals.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> Can this be done as a deprecation of the old "interface" and intro a new Kconfig for the new means and have the various in-trees select that new kconfig?

Adding a Kconfig to preserve the old behavior can be done, I think. I'd not pollute tree with that symbol, though, but offer it to the apps that want to migrate later. But, as I said, despite this being in a public header, the CS field should exclusively be used by drivers, not applications. We should one day isolate better what is driver-facing and application-facing in driver headers because to me, it's not the same to change one side or the other.

gmarull committed:
doc: release-notes: inform about SPI CS changes

Inform about the cs field change in struct spi_config. Also mention what
drivers must use from now on.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
RFC: https://github.com/zephyrproject-rtos/zephyr/issues/57025
Mail list announcement: https://lists.zephyrproject.org/g/devel/message/8940

gmarull commented on drivers/spi/spi_handlers.c:
```diff
@@ -104,8 +104,8 @@ static inline int z_vrfy_spi_transceive(const struct device *dev,
+		const struct spi_cs_control *cs = &config_copy.cs;
 
 		Z_OOPS(Z_SYSCALL_MEMORY_READ(cs, sizeof(*cs)));
 		if (cs->gpio.port) {
```
good catch, fixed
